File: test2/src/main/java/xyz/yggdrazil/delaunay/voronoi/nodename/as3delaunay/HalfedgePriorityQueue.java
package xyz.yggdrazil.delaunay.voronoi.nodename.as3delaunay;

import xyz.yggdrazil.delaunay.geom.Point;

import java.util.ArrayList;

public final class HalfedgePriorityQueue // also known as heap
{

    private ArrayList<Halfedge> _hash;
    private int _count;
    private int _minBucket;
    private int _hashsize;
    private double _ymin;
    private double _deltay;

    public HalfedgePriorityQueue(double ymin, double deltay, int sqrt_nsites) {
        _ymin = ymin;
        _deltay = deltay;
        _hashsize = 4 * sqrt_nsites;
        initialize();
    }

    public void dispose() {
        // get rid of dummies
        for (int i = 0; i < _hashsize; ++i) {
            _hash.get(i).dispose();
        }
        _hash.clear();
        _hash = null;
    }

    private void initialize() {
        int i;

        _count = 0;
        _minBucket = 0;
        _hash = new ArrayList(_hashsize);
        // dummy Halfedge at the top of each hash
        for (i = 0; i < _hashsize; ++i) {
            _hash.add(Halfedge.Companion.createDummy());
            _hash.get(i).setNextInPriorityQueue(null);
        }
    }

    public void insert(Halfedge halfEdge) {
        Halfedge previous, next;
        int insertionBucket = bucket(halfEdge);
        if (insertionBucket < _minBucket) {
            _minBucket = insertionBucket;
        }
        previous = _hash.get(insertionBucket);
        while ((next = previous.getNextInPriorityQueue()) != null
                && (halfEdge.getYstar() > next.getYstar() || (halfEdge.getYstar() == next.getYstar() && halfEdge.getVertex().get_x() > next.getVertex().get_x()))) {
            previous = next;
        }
        halfEdge.setNextInPriorityQueue(previous.getNextInPriorityQueue());
        previous.setNextInPriorityQueue(halfEdge);
        ++_count;
    }

    public void remove(Halfedge halfEdge) {
        Halfedge previous;
        int removalBucket = bucket(halfEdge);

        if (halfEdge.getVertex() != null) {
            previous = _hash.get(removalBucket);
            while (previous.getNextInPriorityQueue() != halfEdge) {
                previous = previous.getNextInPriorityQueue();
            }
            previous.setNextInPriorityQueue(halfEdge.getNextInPriorityQueue());
            _count--;
            halfEdge.setVertex(null);
            halfEdge.setNextInPriorityQueue(null);
            halfEdge.dispose();
        }
    }

    private int bucket(Halfedge halfEdge) {
        int theBucket = (int) ((halfEdge.getYstar() - _ymin) / _deltay * _hashsize);
        if (theBucket < 0) {
            theBucket = 0;
        }
        if (theBucket >= _hashsize) {
            theBucket = _hashsize - 1;
        }
        return theBucket;
    }

    private boolean isEmpty(int bucket) {
        return (_hash.get(bucket).getNextInPriorityQueue() == null);
    }

    /**
     * move _minBucket until it contains an actual Halfedge (not just the dummy
     * at the top);
     */
    private void adjustMinBucket() {
        while (_minBucket < _hashsize - 1 && isEmpty(_minBucket)) {
            ++_minBucket;
        }
    }

    public boolean empty() {
        return _count == 0;
    }

    /**
     * @return coordinates of the Halfedge's vertex in V*, the transformed
     * Voronoi diagram
     */
    public Point min() {
        adjustMinBucket();
        Halfedge answer = _hash.get(_minBucket).getNextInPriorityQueue();
        return new Point(answer.getVertex().get_x(), answer.getYstar());
    }

    /**
     * remove and return the min Halfedge
     *
     * @return
     */
    public Halfedge extractMin() {
        Halfedge answer;

        // get the first real Halfedge in _minBucket
        answer = _hash.get(_minBucket).getNextInPriorityQueue();

        _hash.get(_minBucket).setNextInPriorityQueue(answer.getNextInPriorityQueue());
        _count--;
        answer.setNextInPriorityQueue(null);

        return answer;
    }
}

File: test2/src/main/kotlin/xyz/yggdrazil/fortune/HalfedgePriorityQueue.kt
package xyz.yggdrazil.fortune

import xyz.yggdrazil.math.geometry.Point

import java.util.ArrayList

class HalfedgePriorityQueue // also known as heap
(private val ymin: Double, private val deltay: Double, sqrt_nsites: Int) {

    private var hash: ArrayList<Halfedge>? = null
    private var count: Int = 0
    private var minBucket: Int = 0
    private val hashsize: Int

    init {
        hashsize = 4 * sqrt_nsites

        var i: Int

        count = 0
        minBucket = 0
        hash = ArrayList<Halfedge>(hashsize)
        // dummy Halfedge at the top of each hash
        i = 0
        while (i < hashsize) {
            hash!!.add(Halfedge.createDummy())
            hash!![i].nextInPriorityQueue = null
            ++i
        }
    }

    fun dispose() {
        // get rid of dummies
        for (i in 0..hashsize - 1) {
            hash!![i].dispose()
        }
        hash!!.clear()
        hash = null
    }

    fun insert(halfEdge: Halfedge) {
        var previous: Halfedge
        var next: Halfedge?
        val insertionBucket = bucket(halfEdge)
        if (insertionBucket < minBucket) {
            minBucket = insertionBucket
        }
        previous = hash!![insertionBucket]
        next = previous.nextInPriorityQueue
        while (next is Halfedge && (halfEdge.ystar > next.ystar || halfEdge.ystar == next.ystar && halfEdge.vertex!!._x > next.vertex!!._x)) {
            previous = next
            next = previous.nextInPriorityQueue
        }
        halfEdge.nextInPriorityQueue = previous.nextInPriorityQueue
        previous.nextInPriorityQueue = halfEdge
        ++count
    }

    fun remove(halfEdge: Halfedge) {
        var previous: Halfedge
        val removalBucket = bucket(halfEdge)

        if (halfEdge.vertex != null) {
            previous = hash!![removalBucket]
            while (previous.nextInPriorityQueue != halfEdge) {
                previous = previous.nextInPriorityQueue!!
            }
            previous.nextInPriorityQueue = halfEdge.nextInPriorityQueue
            count--
            halfEdge.vertex = null
            halfEdge.nextInPriorityQueue = null
            halfEdge.dispose()
        }
    }

    private fun bucket(halfEdge: Halfedge): Int {
        var theBucket = ((halfEdge.ystar - ymin) / deltay * hashsize).toInt()
        if (theBucket < 0) {
            theBucket = 0
        }
        if (theBucket >= hashsize) {
            theBucket = hashsize - 1
        }
        return theBucket
    }

    private fun isEmpty(bucket: Int): Boolean {
        return hash!![bucket].nextInPriorityQueue == null
    }

    /**
     * move minBucket until it contains an actual Halfedge (not just the dummy
     * at the top);
     */
    private fun adjustMinBucket() {
        while (minBucket < hashsize - 1 && isEmpty(minBucket)) {
            ++minBucket
        }
    }

    fun empty(): Boolean {
        return count == 0
    }

    /**
     * @return coordinates of the Halfedge's vertex in V*, the transformed
     * * Voronoi diagram
     */
    fun min(): Point {
        adjustMinBucket()
        val answer = hash!![minBucket].nextInPriorityQueue
        return Point(answer!!.vertex!!._x, answer.ystar)
    }

    /**
     * remove and return the min Halfedge

     * @return
     */
    fun extractMin(): Halfedge {
        val answer: Halfedge

        // get the first real Halfedge in minBucket
        answer = hash!![minBucket].nextInPriorityQueue!!

        hash!![minBucket].nextInPriorityQueue = answer.nextInPriorityQueue
        count--
        answer.nextInPriorityQueue = null

        return answer
    }
}